### ISSUE TITLE
iOS build issue

### ISSUE THREAD

**Issue by Bjoe** [randombit/botan#2015]
I try to create a [botan hunter package](https://github.com/ruslo/hunter/issues/1897) but I have some issues to build botan for iOS.

I configure botan for iOS like:
`./configure.py --os=ios --cc=clang --cpu=armv7 ...` [see build on travis line 667](https://travis-ci.org/Bjoe/hunter/jobs/553476103#L667)

Instead of using the `clang` compiler, botan build is using the `g++`  [see build on travis line 704](https://travis-ci.org/Bjoe/hunter/jobs/553476103#L704)

After executing of `make install` [see build on travis line 1102](https://travis-ci.org/Bjoe/hunter/jobs/553476103#L1102)
I get `Unsupported architecture` error.

In the [documentation](https://botan.randombit.net/handbook/building.html#for-ios-using-xcode) I see that the build environment is set via `xcrun` only at build time and not for configure.py. Is this correct?

**Comment by randombit** [comment]
> Instead of using the clang compiler, botan build is using the g++

That definitely looks weird. One way I think this could happen is if CXX is set in the environment, is that possible here?

> I get Unsupported architecture error.

I would guess that's because g++ is emitting x86-64 code instead of ARM code as the iOS SDK compiler would. Probably getting the g++ vs clang situation sorted will resolve this.

> In the documentation I see that the build environment is set via xcrun only at build time and not for configure.py. Is this correct?

Honestly I have no idea. I don't do any mobile dev and all of the docs and code in this area was contributed by others. There may be a problem. I would be happy to accept patches.

**Comment by Bjoe** [comment]
The weird think is that they write [Autodetected platform information: OS="Darwin" machine="x86_64" proc="i386"](https://travis-ci.org/Bjoe/hunter/jobs/553476103#L677) ... looks like they ignore my parameters ?

I will try to set CXX.
Thanks for the hint.

**Comment by randombit** [comment]
> The weird think is that they write Autodetected platform information: OS="Darwin" machine="x86_64" proc="i386" ... looks like they ignore my parameters ?

That log message is misleading it's just logging what the Python `platform` module reports, which is sometimes useful when debugging. Anything given by the user on the command line should be overriding anything that is detected automatically.

**Comment by Bjoe** [comment]
Here some results that I found:
If `CXX` is set .... then ./configure.py take this compiler setting. It doesn't matter if the --cc=clang ... see [for example in the travis build](https://travis-ci.org/Bjoe/hunter/jobs/55411239).
In this build `CXX` [is set to `g++`](https://travis-ci.org/Bjoe/hunter/jobs/554112391#L75)
I set on `configure.py` [the parameter `--cc=clang`](https://travis-ci.org/Bjoe/hunter/jobs/554112391#L669) 
`configure.py` [print out my parameters](https://travis-ci.org/Bjoe/hunter/jobs/554112391#L678) but also [detected a running `g++`](https://travis-ci.org/Bjoe/hunter/jobs/554112391#L682)
As result, in the [Makefile I see `CXX = g++`](https://travis-ci.org/Bjoe/hunter/jobs/554112391#L712)

If I clear the `CXX` environment, then it works. See for example[ in the travis build](https://travis-ci.org/Bjoe/hunter/jobs/554175328).
I [set the `--cc=clang` parameter](https://travis-ci.org/Bjoe/hunter/jobs/554175328#L669)  on `configure.py`.
`configure.py` [print out my setting](https://travis-ci.org/Bjoe/hunter/jobs/554175328#L853) and [detected a running `clang++`](https://travis-ci.org/Bjoe/hunter/jobs/554175328#L857).
Now I see also [in the Makefile `CXX = clang++`](https://travis-ci.org/Bjoe/hunter/jobs/554175328#L887)

Conclusion, `CXX` overules `--cpu=` setting, correct?

**Comment by Bjoe** [comment]
> One way I think this could happen is if CXX is set in the environment, is that possible here?

Ok now I got it .... it is a default behavior ... but again, a warning would be nice because I don't now if it make sense when a "not seen" `CXX` environment is set additional to an `--cpu=` parameter .... it looks like not intuitive for me

**Comment by randombit** [comment]
The command line is preferred over the environment, but there are two different options involved here, one is `--cc` which is "what kind of compiler is this (say, gcc)" vs `--cc-bin` is "what is the path to invoke this compiler (say, /opt/bin/aarch64-linux-gnu-g++-8.3)". If `--cc-bin` is not set then the value of `CXX` is used.

You're right this is confusing, when you don't know that CXX is set in the first place. I'll add some logging so it's easier to spot.

**Comment by randombit** [comment]
Now we log when this happens. And it sounds like you have resolved your build problem. Can this issue be closed?

**Comment by Bjoe** [comment]
Unfortunately I get now a new build problem: [error: Unsupported architecture](https://travis-ci.org/Bjoe/hunter/jobs/554204626#L894)

I will take a Mac with an iPhone SDK installed to debug on a real machine.
I can maybe close this issue for that time.... or maybe somebody has an idea what can be now wrong?

**Comment by randombit** [comment]
I think the issue there is how you are quoting "-arch arm64"

"--cc-abi-flags=\"-arch arm64\""

This ends up being passed as a single argument to Clang which doesn't understand it.

clang: warning: argument unused during compilation: '-arch arm64' [-Wunused-command-line-argument]

I think you want just a single set of quotes around the configure arg "--cc-abi-flags=-arch arm64"

**Comment by Bjoe** [comment]
> I think the issue there is how you are quoting "-arch arm64"
@randombit excellent hint! It compiles now but .....

Next step next level ;-) .... puh compiler crash ... In function: _ZN5Botan15newhope_sharedbEPhS0_PKhRNS_21RandomNumberGeneratorENS_12Newhope_ModeE [fatal error: error in backend: Cannot select: 0x7f8ccad00740: v16i8 = insert_subvector undef:v16i8, 0x7f8ccb01b018, Constant:i32<0>](https://travis-ci.org/Bjoe/hunter/jobs/554627173#L2549) 

Some idea? Compiler to old?

**Comment by randombit** [comment]
Ah yes. We ran into this in CI as well, and for that reason the 32-bit iOS CI build is disabled. This file compiled fine with older versions of XCode. It may work with newer versions, but I haven't tried. It doesn't affect 64-bit build. I tried to report the bug at some point but ISTR doing so required an Apple Developer ID and I'm not paying them just to tell them they have a buggy compiler.

I guess, try latest XCode. If that doesn't work, then use `--disable-modules=newhope` for 32-bit.
